### DISCUSSION TITLE
How to handle authentication in Nextjs

### DISCUSSION THREAD

**Discussion by kaisiant** [vercel/next.js#19917] [Help]
**Question** 
Since I don't want to show the loading state, I need to call the api in getServerSideProps, but the problem is JWT is storing in the local storage and using ssr is not possible to get it. Should I change the api to use cookie? Or **js-cookie** might help? What is the correct way to implement the authentication?

The current way of authenticating user in my project.
Problem: keep showing loading on refresh
```
export default function ({ Component, pageProps, router }) {
  const { data, isLoading } = useMe();
  if (isLoading) return "loading";
  const isAdmin = data?.customers?.[0].role_ids?.includes(1);

  const C = <Component {...pageProps} />;
  const renderComponent = () => {
    if (router.pathname.startsWith("/admin") && isAdmin) {
      return <AdminLayout>{C}</AdminLayout>;
    }

    return <BasicLayout>{C}</BasicLayout>;
  };
  return (
    <ReactQueryCacheProvider queryCache={queryCache}>
      <Hydrate state={pageProps.dehydratedState}>{renderComponent()}</Hydrate>
    </ReactQueryCacheProvider>
  );
}

```

**Comment by mnlfischer** [comment]
Hello @kaisian97, 
I had exactly the same question and tried to implement different methods.
JWT was not an option through XSS (and there is no option to get it on server side). I dont want to check all libs I use if they are XSS protected, so my advice is to use cookies.
I tried [magic link](https://magic.link/), but they don't provide any mechanism to delete user relevant data, so it is not safe to use in EU (laws).
In my opinion [nextauth](https://next-auth.js.org/) is a good solution with everything what you need. You can get the users session on serverside, api routes, client side and had a built in support for csrf protection. Please check supported databases before use, but you can also create your own database adapter if you have something special.
